pkg/client/history/table.go

```diff
@@ -0,0 +1,144 @@
+	"github.com/georgysavva/scany/pgxscan"
+	"github.com/hashicorp/go-hclog"
+	"github.com/huandu/go-sqlbuilder"
+	"github.com/jackc/pgx/v4/pgxpool"
+)
+
+type CreateHyperTableResult struct {
+	HypertableId int    `db:"hypertable_id"`
+	SchemaName   string `db:"schema_name"`
+	TableName    string `db:"table_name"`
+	Created      bool   `db:"created"`
+}
+
+type TableCreator struct {
+	log hclog.Logger
+	cfg *Config
+}
+
+func NewHistoryTableCreator(cfg *Config, l hclog.Logger) (*TableCreator, error) {
+	return &TableCreator{
+		l,
+		cfg,
+	}, nil
+}
+
+func (h TableCreator) CreateTable(ctx context.Context, conn *pgxpool.Conn, t *schema.Table, p *schema.Table) error {
+	sql, err := h.buildTableSQL(t, p)
+	if err != nil {
+		return err
+	}
+	h.log.Debug("creating table if not exists", "table", t.Name)
+	if _, err := conn.Exec(ctx, sql); err != nil {
+		return fmt.Errorf("failed to create table: %w", err)
+	}
+
+	if err := h.createHyperTable(ctx, t, conn); err != nil {
+		return fmt.Errorf("failed to created hypertable for table: %s: %w", t.Name, err)
+	}
+
+	if _, err := conn.Exec(ctx, fmt.Sprintf(createTableView, t.Name)); err != nil {
+		return fmt.Errorf("failed to created view for table: %s: %w", t.Name, err)
+	}
+
+	if p != nil {
+		if err := h.buildCascadeTrigger(ctx, conn, t, p); err != nil {
+			return fmt.Errorf("table build %s failed: %w", t.Name, err)
+		}
+	}
+
+	// Create relation tables
+	for _, r := range t.Relations {
+		h.log.Debug("creating table relation", "table", r.Name)
+		if err := h.CreateTable(ctx, conn, r, t); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (h TableCreator) createHyperTable(ctx context.Context, t *schema.Table, conn *pgxpool.Conn) error {
+	var hyperTable CreateHyperTableResult
+	tName := fmt.Sprintf(`"history"."%s"`, t.Name)
+	if err := pgxscan.Get(ctx, conn, &hyperTable, fmt.Sprintf(createHyperTable, h.cfg.TimeInterval), tName); err != nil {
+		return fmt.Errorf("failed to create hypertable: %w", err)
+	}
+	h.log.Debug("created hyper table for table", "table", hyperTable.TableName, "id", hyperTable.HypertableId, "created", hyperTable.Created)
+	if _, err := conn.Exec(ctx, fmt.Sprintf(dataRetentionPolicy, h.cfg.Retention), tName); err != nil {
+		return err
+	}
+	return nil
+}
+
+// TODO: create unique index on fetch_date parent_cq_id
+func (h TableCreator) buildCascadeTrigger(ctx context.Context, conn *pgxpool.Conn, t *schema.Table, p *schema.Table) error {
+	c := h.findParentIdColumn(t)
+	if c == nil {
+		return fmt.Errorf("failed to find parent cq id column for %s", t.Name)
+	}
+	_, err := conn.Exec(ctx, "SELECT history.build_trigger($1, $2, $3);", p.Name, t.Name, c.Name)
+	if err != nil {
+		return fmt.Errorf("failed to create trigger: %w", err)
+	}
+	return nil
+}
+
+func (h TableCreator) findParentIdColumn(t *schema.Table) *schema.Column {
+	for _, c := range t.Columns {
+		if strings.HasSuffix(c.Name, "cq_id") && c.Name != "cq_id" {
```

Comment:
Looks like a polyfill for older providers that don't support `c.Meta()`, so maybe move this after the Meta check and add comment to clarify?

Also, maybe go over `t.Columns` two times, so that if the `foo_cq_id` isn't the parent reference but meta is supported, we don't end up choosing the wrong column just based on the name (first one we encounter basically)